File: src/libs/cpp/functions/random_weighted.cpp
#include "random_weighted.hpp"
#include <iostream>

void random_weighted(Context &context) noexcept
{
	auto args = std::get<std::vector<Value>>(context.stack.pop());
	auto array = args[0].to_array();
	auto weights = args[1].to_array();

	const auto length = std::min(array.size(), weights.size());

	// Calculate the total weight
	double total_weight = 0;
	for (const auto &weight : weights)
	{
		total_weight += weight.to_number();
	}

	double r = ((double)(context.rng() % RAND_MAX) / RAND_MAX) * total_weight;
	double cumulative = 0;

	for (size_t i = 0; i < length; i++)
	{
		cumulative += weights[i].to_number();
		std::cout << r << " " << cumulative << std::endl;
		if (r < cumulative)
		{
			context.stack.push(array[i]);
			return;
		}
	}

	context.stack.push(array.back());
}

File: src/runtime/functions/random_weighted.lua
return function(vm)
	local v = vm.pop()
	local vector, weights = v[1], v[2]
	if std.type(vector) ~= 'array' then
		vector = { std.num(vector) }
	end
	if std.type(weights) ~= 'array' then
		weights = { std.num(weights) }
	end
	vm.push(std.random_weighted(vector, weights))
end
